Change the editor links when Dashboard Appearance toggle is enabled

pablinos opened the pull request:
The links to the editor still default to the iframe'd `/posts` URLs and
that means that people can find themselves back in Calypso even though
they have enabled the 'Dashboard Appearance' option, which should
default to `wp-admin`.



#### Changes proposed in this Pull Request

This change, updates the logic in the selector for these links, as well
as the redirect logic as the iframe loads, to take the 'Dashboard
Appearance' setting in to account.

Fixes https://github.com/Automattic/wp-calypso/issues/51058

#### Testing instructions

As outlined in #51058.

* Change your 'Dashboard Appearance' setting in `/me/account` so that it is enabled.
* Click on the 'Write' button in the top bar. This should now take you to the `wp-admin` version of the editor.
* Clicking on the 'All Posts' link in the left side nav, should take you back to `wp-admin`
* Also check that heading to /post/<YOUR SITE> redirects you to `wp-admin`


<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

#### Notes

It's worth noting that this is relying on something that could be considered a bug. When the editor is loaded outside of the iframe, the code to override the editor `closeUrl` isn't loaded, so it is left as the default, which is `wp-admin`. By making sure that we load the editor in `wp-admin`, outside of the iframe, then the link works as expected. 

We should change this so the code takes the 'Dashboard Appearance' setting into account and loads the code when appropriate. This is a much larger change though.

Also, I think there will need to be a separate change to update the link in the top navigation bar in `wp-admin`, as well as possibly the WordPress.com navigation module in Jetpack.

Related to #51058

pablinos committed:
Change the editor links when Dashboard Appearance toggle is enabled

The links to the editor still default to the iframe'd `/posts` URLs and
that means that people can find themselves back in Calypso even though
they have enabled the 'Dashboard Appearance' option, which should
default to `wp-admin`.

This change, updates the logic in the selector for these links, as well
as the redirect logic as the iframe loads, to take the 'Dashboard
Appearance' setting in to account.

allilevine commented:
> 
>     * Change your 'Dashboard Appearance' setting in `/me/account` so that it is enabled.
> 
>     * Click on the 'Write' button in the top bar. This should now take you to the `wp-admin` version of the editor.
> 
>     * Clicking on the 'All Posts' link in the left side nav, should take you back to `wp-admin`
> 
>     * Also check that heading to /post/ redirects you to `wp-admin`

All of the links took me to `wp-admin` except `/post/`. 🤔 I'm testing by going to `http://calypso.localhost:3000/post/blog-url`.

pablinos commented:
> All of the links took me to wp-admin except /post/. 🤔 I'm testing by going to http://calypso.localhost:3000/post/blog-url.

Thanks! Sounds like something isn't right, and the tests are failing. I'll take another look.

pablinos commented:
So, after some investigation, it appears that what is happening is that the `calypso_preferences` in the state are `undefined` until the API request returns successfully. It's a race condition where sometimes this comes back before the redirect check and sometimes it doesn't.

I'm not sure of the best way to solve this, as we need the preference to know whether to redirect, and we could wait for it in the state, but I've had a situation where that request comes back as unauthorised, so we'd need to deal with the error state too. That feels like a lot of logic to live in the controller. 

I feel like I'm missing something obvious. I'll give it some more thought.

mmtr commented:
> it appears that what is happening is that the calypso_preferences in the state are undefined until the API request returns successfully

I remember we had a similar issue when we first introduced Gutenberg in Calypso a couple of years ago, since we needed to retrieve data from the API before deciding the editor URL. We solved that by:
- Querying the needed data in the `Layout` component: 
https://github.com/Automattic/wp-calypso/blob/f8531073cb6c5f77b4351569426872c3b3afa5cd/client/layout/index.jsx#L235-L237
- Ensuring the data has been received in the editor routes controller:
https://github.com/Automattic/wp-calypso/blob/f8531073cb6c5f77b4351569426872c3b3afa5cd/client/gutenberg/editor/controller.js#L175-L178

pablinos commented:
Thanks, @mmtr, I was wondering if I needed to do that. I started writing a similar `waitForDesktopAppearanceSetting` but got into a funny state where the API call was returning an error. In those situations, I think things would get stuck. I'll see if there's a simple way to check for errors and default one way or the other.

It also seemed like a complex solution and so I wondered if I was missing something.

pablinos committed:
Wait for the Calypso preferences in the editor controller

This adds a function to wait for the Calypso preferences in the editor
controller. That way we can determine the state of the dashboard
appearance setting before deciding whether to redirect to wp-admin or
load the Gutenframe.

pablinos committed:
Fix unit test.

pablinos committed:
Fix the unit tests that touch isNavUnificationEnabled

pablinos committed:
An attempt to avoid the use of document to fix SSR

retrofox reviewed (commented):
I've tested the app activating and deactivating the "Dashboard appearance". To works as expected.
The links when creating a new post link points to wp-admin when "Dashboard appearance" is activated. If I type manually the URL to start to edit using the iframe, the app redirects on the fly to wp-admin too.

I have a question about the posts list page. The link for every post seems to follow the _Dashboard appearance_ rule. It means, when it's activated, the URL points to the wp-admin. Otherwise, it points to the calypso app: 

<img width="851" alt="Screen Shot 2021-03-29 at 3 34 14 PM" src="https://user-images.githubusercontent.com/77539/112883471-b7f11b00-90a4-11eb-9628-875d0b774a38.png">

However, the `All posts` link in the uni-nav menu always points to the wp-admin, independently of the _Dashboard appearance_ state.

<img width="884" alt="Screen Shot 2021-03-29 at 3 33 07 PM" src="https://user-images.githubusercontent.com/77539/112883431-a9a2ff00-90a4-11eb-92c8-e7e1aa42ab8a.png">

Is it something to deal with?

pablinos commented:
Hmmm, that seems like a bug. I'll check if it's introduced by this change and if not then I think we should look at it separately.

retrofox commented:
> Hmmm, that seems like a bug. I'll check if it's introduced by this change and if not then I think we should look at it separately.

Let me help with this too.

retrofox commented:
it seems to happen even in production too

retrofox reviewed (approved):
Going to approve it since what I commented before seems to happen in prod too.

pablinos commented:
Yes, I think the site(s) you're seeing this for are Jetpack sites, where wp-admin is the default for all nav links. Simple sites that I've tested this on, the link does change based on the setting

pablinos commented:
@mmtr Would you mind taking a look at this when you've got a chance please? I'm waiting for both the selected editor and the user preferences to be available before continuing with the routing logic in the controller. I've tried to cover the error condition where we just continue - things should get picked up later, it's not the end of the world if someone ends up in the wrong place and if we can't get the user settings for some reason then it's a sign of something more serious.

Can you see any problems with this? I don't want people getting stuck in this redirect check.

mmtr commented on client/state/selectors/should-load-gutenframe.js:
```diff
@@ -0,0 +1,17 @@
+export const shouldLoadGutenframe = ( state, siteId ) => {
+	return (
+		isEligibleForGutenframe( state, siteId ) &&
+		shouldCalypsoifyJetpack( state, siteId ) &&
```
Nitpick: `isEligibleForGutenframe` already checks `shouldCalypsoifyJetpack`, so do we need to check it again here?

pablinos replied:
Yes, that's true. I think I must have forgotten we had it in there. I've updated it.

mmtr reviewed (approved):
The proposed approach looks good to me. Nice job, @pablinos.

mmtr commented:
However, I think this will also need a counterpart fix for WP Admin, so the "Write" button of the top bar displayed on WP Admin screens also links to the same URL.

pablinos commented:
> However, I think this will also need a counterpart fix for WP Admin, so the "Write" button of the top bar displayed on WP Admin screens also links to the same URL.

Yes, I've been looking at this. I think we've so far taken the approach of sending everyone to `/post` from `wp-admin`. I'll need to look the best way to replicate the logic.

pablinos committed:
Remove unnecessary call to shouldLoadGutenframe

noahtallen commented:
I this introduced a performance regression, see: https://github.com/Automattic/wp-calypso/issues/52651
